forward_service: introduce shutdown checks

havaker opened the pull request:
This commit introduces a new boolean flag, `shutdown`, to the `forward_service`, along with a corresponding `shutdown` method. It also adds checks throughout the `forward_service` to verify the value of the `shutdown` flag before retrying or invoking functions that might use the messaging service under the hood.

The flag is set before messaging service shutdown, by invoking `forward_service::shutdown` in main. By checking the flag before each call that potentially involves the messaging service, we can ensure that the messaging service is still operational. If the flag is false, indicating that the messaging service is still active, we can proceed with the call. In the event that the messaging service is shutdown during the call, appropriate exceptions should be thrown somewhere down in called functions, avoiding potential hangs.

This fix should resolve the issue where `forward_service` retries could block the shutdown.

Fixes #12604

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1299/

havaker commented on service/forward_service.cc:
```diff
@@ -277,6 +281,10 @@ class retrying_dispatcher {
                 &_forwarder._messaging, id, req, _tr_info
             ).handle_exception_type([this, &req, &id] (rpc::closed_error& e) -> future<query::forward_result> {
+                if (_forwarder._shutdown) {
+                    throw;
```
Oops, honestly, I'm not really sure. I'll dig into it and do some research to figure out if it's right.

havaker replied:
`handle_exception_type` invokes its lambda parameter using `futurize_invoke`, which catches thrown exceptions and returns `current_exception_as_future()`. So this code should be correct and return an exceptional future afterall.

This rethrow needed an explanation so it's better to replace it with something simpler - explicit returning of an exceptional future.

havaker commented:
v2:

- replaced rethrow with explicit return of an exceptional future

xemul commented on service/forward_service.cc:
```diff
@@ -266,6 +266,10 @@ class retrying_dispatcher {
 
     future<query::forward_result> dispatch_to_node(netw::msg_addr id, query::forward_request req) {
+        if (_forwarder._shutdown) {
+            throw std::runtime_error("forward_service is shutting down");
```
My previous comment regarding throw vs make_exception_future<> applies here too

xemul commented on service/forward_service.cc:
```diff
@@ -462,6 +476,10 @@ future<query::forward_result> forward_service::execute_on_this_shard(
             co_await pager->fetch_page(rs_builder, DEFAULT_INTERNAL_PAGING_SIZE, now, timeout);
+
+            if (_shutdown) {
+                throw std::runtime_error("forward_service is shutting down");
```
And here

havaker replied:
Why? I thought that your previous comment pointed out that throwing in a non-coroutine lambda may have surprising effects. This method is a coroutine, could you elaborate on what is wrong here?

xemul commented on service/forward_service.cc:
```diff
@@ -497,6 +515,9 @@ void forward_service::init_messaging_service() {
+            if (_shutdown) {
+                throw std::runtime_error("forward_service is shutting down");
+            }
             return dispatch_to_shards(req, tr_info);
```
It looks like you put `if (_sutdown)` checks in front of every invocation of `dispatch_to_shards()` and not more than that. Maybe put a single if into this method only?

havaker replied:
I have adjusted the positioning of shutdown checks. They now precede the calls to functions that may potentially utilize the messaging service, namely `pager->fetch_page()` and `ser::forward_request_rpc_verbs::send_forward_request()`.

This should hopefully fix the issue of `forward_service` hanging on rpc calls after messaging service shutdown.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1332/

havaker commented:
v3:

- replaced throwing with explicit returnal of an exceptional future
- [repositioned](https://github.com/scylladb/scylladb/pull/13922#discussion_r1227352451) shutdown checks
- added comments that explain motivation behind position of each shutdown check

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1785/
